### PR TITLE
ci(GHA): Improve handling of PRs from forks

### LINKED ISSUE #2737: Refactor CI workflow to allow PR's from forks

### PR DESCRIPTION
**Pull request by jpveooys**
## Related issue

Resolves #2737 

## Overview

This:

- makes Chromatic work with PRs from forks
- gets Chromatic reporting its status in GitHub properly again
- adds a GitHub code security scan for all PRs
- gracefully skips SonarCloud for PRs from forks (due to incompatibility)
- removes the use of secrets in cache keys
- updates all actions to their latest versions

## Reason

To make sure PRs can be made from forks without friction.

## Work carried out

- [x] Update all actions to latest versions
- [x] Reformat all workflows
- [x] Remove use of secret in cache keys
- [x] Skip SonarCloud scan on PRs from forks
- [x] Add GitHub code scanning
- [x] Handle Chromatic builds in PRs from forks

## Developer notes

See https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ for some detail on the relevant considerations here.

This has been tested on a fork:

- [example of local PR](https://github.com/jpveooys/github-actions-testing/pull/2)
- [example of PR from a fork with visual changes](https://github.com/jpveooys/github-actions-testing/pull/5)
- [example of automerged PR](https://github.com/jpveooys/github-actions-testing/pull/8)
- [example of unmerged major dependency bump PR](https://github.com/jpveooys/github-actions-testing/pull/6)

Chromatic has been refactored to follow a similar pattern as in the blog post linked above (for all builds). This means Storybook is built in the main PR build and uploaded as an artefact, and then this is downloaded in an `on: workflow_run` workflow where master is checked out instead of the PR branch and secrets are available.

This also brings back the Chromatic status on PRs:

image

image

These won't appear until the PR is merged (as  `.github/workflows/post_build_and_test.yml` needs to be on master for it to be used).

**Branch protection rules will need to be updated after this has been merged to require UI Tests instead of Test_visual_regression.**

There's no nice way to make SonarCloud work with forks, because you have to have the secret available and the code checked out at the same time (this could be done, but it's not generally recommended). They have support for forks on their product board: https://portal.productboard.com/sonarsource/1-sonarcloud/c/50-sonarcloud-analyzes-external-pull-request

I added [GitHub code scanning](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning) to make sure PRs from forks do go through a code security scan.

### CODE REVIEW

**Review by thyhjwb6 — APPROVED** [review]
Nice 👍

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -12,16 +12,16 @@ jobs:
...
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
 
       - name: Cache Node modules
-        uses: actions/cache@v2
+        uses: actions/cache@v3
         with:
           path: '**/node_modules'
-          key: ${{ runner.os }}-modules-${{ secrets.CACHE_VERSION }}-${{ hashFiles('**/yarn.lock') }}
+          key: ${{ runner.os }}-node-modules-${{ hashFiles('**/yarn.lock') }}
```
Comment (m7kvqbe1) [inline .github/workflows/build_and_test.yml:21]: What was the reason for that `CACHE_VERSION` secret? 

Was it there as a bit of a hack to manually invalidate the cache?

File: .github/workflows/post_build_and_test.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        run: |
+          yarn install --frozen-lockfile
+
+      # Based on
+      # https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#using-data-from-the-triggering-workflow
+      - name: Download pre-built Storybook
+        uses: actions/github-script@v6
+        with:
+          script: |
+            const allArtifacts = await github.rest.actions.listWorkflowRunArtifacts({
```
Comment (m7kvqbe1) [inline .github/workflows/post_build_and_test.yml]: Is it worth putting this in an external seperate script and referencing it?

**Review by m7kvqbe1 — COMMENTED** [review]
Nice changes! Just a couple of small comments.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (jpveooys) [reply .github/workflows/build_and_test.yml:21]: Yes that's probably the main reason. The other possibility is changing it after e.g. a Node.js upgrade.

We could keep it in but it won't work for forks or Dependabot. We could also just append e.g. v1 to the cache key and change it in the workflow if a problem came up.

It seems like some APIs to manage the cache may be coming later in the year as well: https://github.com/actions/cache/issues/2#issuecomment-1014164408

File: .github/workflows/post_build_and_test.yml (hunk 2 above)
Comment (jpveooys) [reply .github/workflows/post_build_and_test.yml]: Hmm, yes, it was a bit annoying editing that here. It looks like it's possible: https://github.com/actions/github-script#run-a-separate-file

Will give it a go :+1:

Comment (jpveooys) [reply .github/workflows/post_build_and_test.yml]: Got this working – have pushed a fixup.

**Review by m7kvqbe1 — APPROVED** [review]

**Pull request merged**